### PR TITLE
Add ROCm GPU support for cpp codegen

### PR DESCRIPTION
**Pull request by jataylo**
Adds support for hsaco binaries in inductor's cpp_wrapper codegen.

### CODE REVIEW

**Comment by jataylo** [comment]
Looking for a general review before we upstream this

File: torch/_inductor/codegen/wrapper.py [hunk 1]
```diff
@@ -85,10 +86,16 @@ class EnterCudaDeviceContextManagerLine:
     def codegen(self, code: IndentedBuffer, device_cm_stack: contextlib.ExitStack):
         if V.graph.cpp_wrapper:
             code.writeline("\n")
-            if self.first_time:
-                code.writeline(f"at::cuda::CUDAGuard device_guard({self.device_idx});")
+            if torch.version.hip is None:
+                if self.first_time:
+                    code.writeline(f"at::cuda::CUDAGuard device_guard({self.device_idx});")
+                else:
+                    code.writeline(f"device_guard.set_index({self.device_idx});")
```
Comment (dllehr-amd) [inline torch/_inductor/codegen/wrapper.py:93]: The device guard looks to be the same for both hip and cuda?  could we just have if self.first_time: then check for hip vs cuda?

Comment (jataylo) [reply torch/_inductor/codegen/wrapper.py:93]: Ah thanks for spotting that, this is a mistake that I brought in when copying to this branch.

We need to use HIPGuard
`code.writeline(f"at::hip::HIPGuard device_guard({self.device_idx});")`

**Review by jeffdaily — APPROVED** [review]

**Comment by jataylo** [comment]
Opened up the upstream PR: https://github.com/pytorch/pytorch/pull/101797

**Pull request closed without merging**